src/util/CleanupDeleter.h

```diff
@@ -0,0 +1,41 @@
+//   Copyright 2023, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#ifndef QLEVER_CLEANUPDELETER_H
+#define QLEVER_CLEANUPDELETER_H
+
+#include <functional>
+#include <memory>
+
+namespace ad_utility::cleanup_deleter {
+
+/// Special type of deleter class that allows to call a function
+/// just before the wrapper value T is deleted.
+template <typename T>
+class CleanupDeleter {
+  using type = std::remove_reference_t<T>;
+
+  std::function<void(type&)> function_;
+
+  explicit CleanupDeleter(std::function<void(type&)> function)
+      : function_{std::move(function)} {}
+
+ public:
+  /// Calls function_ and deletes the value afterwards.
+  void operator()(T* ptr) const {
+    function_(*ptr);
+    delete ptr;
+  }
+
+  /// Creates a std::unique_ptr that calls the passed function right before
+  /// obj is about to be deleted.
+  static std::unique_ptr<T, CleanupDeleter<T>> cleanUpAfterUse(
+      std::remove_reference_t<T> obj, std::function<void(type&)> function) {
+    return {new T(std::move(obj)), CleanupDeleter{std::move(function)}};
```

Comment:
@joka921 While `std::unique_ptr` provides a nice part of this kind of interface it still requires a lot of boilerplate code to make it work, so I'm starting to believe a custom solution using a boolean might be the cleaner approach for this, while still keeping the abstraction layer. Thoughts?